sharedbuffers/mapped_struct.py

```diff
@@ -514,27 +514,29 @@ def get_wrapped_key(obj):
 class mapped_frozenset(frozenset):
 
     @classmethod
-    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, cdtype=cython.uchar)
+    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, py2_dtype=cython.uchar, py3_dtype="unicode")
     def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
         if isinstance(obj, npndarray):
             all_int = all_intlong = all_float = 0
             obj_dtype = obj.dtype
             if obj_dtype.isbuiltin:
                 if python3:
-                    dtype = obj_dtype.char.encode()
-                    if cython.compiled:
-                        cdtype = dtype[0]
-                    else:
-                        cdtype = bytes([dtype[0]])
+                    py3_dtype = obj_dtype.char
+                    if py3_dtype[0] in (u'l', u'I', u'i', u'H', u'h', u'B', u'b'):
+                        all_int = all_intlong = 1
+                    elif py3_dtype[0] == u'L':
+                        all_intlong = 1
+                    elif py3_dtype[0] in (u'd', u'f'):
+                        all_float = 1
                 else:
-                    dtype = obj_dtype.char
-                    cdtype = cython.cast('const char*', dtype)[0]
-                if cdtype in (b'l', b'I', b'i', b'H', b'h', b'B', b'b'):
-                    all_int = all_intlong = 1
-                elif cdtype == b'L':
-                    all_intlong = 1
-                elif cdtype in (b'd', b'f'):
-                    all_float = 1
+                    temp = obj_dtype.char
+                    py2_dtype = cython.cast('const char*', temp)[0]
```

Comment:
Add a comment stating why you need temp (to keep a reference to the string while you access the char*)

Otherwise I can totally see someone thinking it's not necessary and causing issues later on